arraycontext/pytest.py

```diff
@@ -28,48 +29,96 @@
 THE SOFTWARE.
 """
 
+from typing import Optional, Sequence
 
-# {{{ pytest integration
+import pyopencl as cl
+from pyopencl.tools import _ContextFactory
 
-def pytest_generate_tests_for_pyopencl_array_context(metafunc):
-    """Parametrize tests for pytest to use a
-    :class:`~arraycontext.PyOpenCLArrayContext`.
 
-    Performs device enumeration analogously to
-    :func:`pyopencl.tools.pytest_generate_tests_for_pyopencl`.
+# {{{ array context factories
 
-    Using the line:
+class _PyOpenCLArrayContextFactory(_ContextFactory):
+    force_device_scalars = True
+
+    def __call__(self):
+        ctx = super().__call__()
+        from arraycontext.impl.pyopencl import PyOpenCLArrayContext
+        return PyOpenCLArrayContext(
+                cl.CommandQueue(ctx),
+                force_device_scalars=self.force_device_scalars)
+
+    def __str__(self):
+        return ("<array context factory for <pyopencl.Device '%s' on '%s'>" %
+                (self.device.name.strip(),
+                 self.device.platform.name.strip()))
+
+
+class _DeprecatedPyOpenCLArrayContextFactory(_PyOpenCLArrayContextFactory):
+    force_device_scalars = False
+
+
+_ARRAY_CONTEXT_FACTORY_DICT = {
+        "pyopencl": _PyOpenCLArrayContextFactory,
+        }
+_ALL_ARRAY_CONTEXT_FACTORY_DICT = {
+        "pyopencl-deprecated": _DeprecatedPyOpenCLArrayContextFactory,
+        }
+
+_ALL_ARRAY_CONTEXT_FACTORY_DICT.update(_ARRAY_CONTEXT_FACTORY_DICT)
+
+# }}}
+
+
+# {{{ pytest integration
+
+def pytest_generate_tests_for_array_context(
+        metafunc,
+        impls: Optional[Sequence[str]] = None):
```

Comment:
I like this, however I would prefer the type to be `Optional[Sequence[Union[str, ArrayContext]]]`. A use case for this would be say meshmode decides to go with its own `ArrayContext` implementation for transformations that are specific to it, then they would want to add that in here.

Reply:
That wouldn't work because it needs to be able to instantiate an `ArrayContext` from a `cl.Device` (at least the way it's currently all set up). How about adding an undocumented
```
def register_array_context_factory(
        name: str,
        func: Callable[[Any], Callable[[], ArrayContext]]) -> None:
```
? Too complicated? I quite like the string-only bit because you can select them from the command line.

Reply:
Also responding to https://github.com/alexfikl/arraycontext/pull/1#discussion_r657513197:

I'm OK with allowing strings for brevity and CLI access, but I think I would like custom, user provided factories to also be allowed. (`register_...` ultimately is based on global shared state, it could lead to name clashes, ...)

Reply:
To respond here too, documented a barebones interface for user provided factories in a77105e. This should allow adding pyopencl-friendly array contexts that we need for the moment.